include/deal.II/grid/grid_generator.h

```diff
@@ -2360,6 +2364,143 @@ namespace GridGenerator
+    using value_type = typename VectorType::value_type;
+
+    /**
+     * Constructor.
+     */
+    MarchingCubeAlgorithm(const Mapping<dim, dim> &      mapping,
+                          const FiniteElement<dim, dim> &fe,
+                          const unsigned int             n_subdivisions = 1);
+
+    /**
+     * Process all locally-owned cells.
+     */
+    void
+    process(const DoFHandler<dim> &         background_dof_handler,
+            const VectorType &              ls_vector,
+            const double                    iso_level,
+            std::vector<Point<dim>> &       vertices,
+            std::vector<CellData<dim - 1>> &cells) const;
+
+    /**
+     * Process a cell (2D).
+     *
+     * @note Subcells with saddle points are ignored. Please increase the number
+     *   of subdivisions in this case.
+     */
+    void
+    process_cell(std::vector<value_type> &    ls_values,
+                 const std::vector<Point<2>> &points,
+                 const double                 iso_level,
+                 std::vector<Point<2>> &      vertices,
+                 std::vector<CellData<1>> &   cells) const;
+
+    /**
+     * Process a cell (3D).
+     */
+    void
+    process_cell(std::vector<value_type> &    ls_values,
+                 const std::vector<Point<3>> &points,
+                 const double                 iso_level,
+                 std::vector<Point<3>> &      vertices,
+                 std::vector<CellData<2>> &   cells) const;
+
+    /**
+     * Process the cell @p cell. Calls the above functions but allows the users
+     * to perform their own cell loop and potentially perform some pre- and
+     * post-processing steps.
+     */
+    void
+    process_cell(const typename DoFHandler<dim>::active_cell_iterator &cell,
+                 const VectorType &              ls_vector,
+                 const double                    iso_level,
+                 std::vector<Point<dim>> &       vertices,
+                 std::vector<CellData<dim - 1>> &cells) const;
+
+  private:
+    /**
+     * Internal function to create a quadrature rule with n_subdivisions+1
+     * equally-positioned quadrature points.
+     */
+    static Quadrature<dim>
+    create_qudrature_rule(const unsigned int n_subdivisions);
+
+    /**
+     * Process a sub-cell (2D).
+     *
+     * @note Subcells with saddle points are ignored. Please increase the number
+     *   of subdivisions in this case.
+     */
+    static void
+    process_sub_cell(const std::vector<value_type> & ls_values,
+                     const std::vector<Point<2>> &   points,
+                     const std::vector<unsigned int> mask,
+                     const double                    iso_level,
+                     std::vector<Point<2>> &         vertices,
+                     std::vector<CellData<1>> &      cells);
+
+    /**
+     * Process a sub-cell (3D).
+     */
+    static void
+    process_sub_cell(const std::vector<value_type> & ls_values,
```

Comment:
Why can't this be templated on `dim`?

Reply:
@Rombur Thanks for the review till now. I have incorporated the changes. However, I would like to keep this function as it is. It is private and my fear is that (as long as we don't support if constexpr) we will get warnings which would have to "fixed" in an other way.